python/rikai/parquet/dataset.py

```diff
@@ -140,13 +151,35 @@ def _convert(
         return converted
 
     def __iter__(self):
+        shuffler = RandomShuffler(
+            self.shuffler_capacity if self.shuffle else 1, self.seed
+        )
+        group_count = 0
         for filepath in self.files:
             fs, path = FileSystem.from_uri(filepath)
             with fs.open_input_file(path) as fobj:
                 parquet = pg.ParquetFile(fobj)
                 for group_idx in range(parquet.num_row_groups):
+                    # A simple form of row-group level bucketing without memory overhead.
+                    # Pros:
+                    #  - It requires zero communication to initialize the distributed policy
+                    #  - It uses little memory and no startup overhead, i.e. collecting row groups.
+                    # Cons:
+                    #   The drawback would be if the world size is much larger than
+                    #   the average number of row groups. As a result, many of the
+                    #   file open operations would be wasted.
+                    group_count += 1
+                    if group_count % self.world_size != self.rank:
```

Comment:
it would not. `parquet.num_row_groups` is different from file to file. Also `num_row_groups` of each file might be a number that is smaller than `world_size` .

This line is basically equivalent to scan all files to get all row groups, and partition them.